File: src/SerialToKeyboardTest/src/main/java/io/github/chase22/keyboardserial/serial/FcReader.java
package io.github.chase22.keyboardserial.serial;

import jssc.SerialPort;
import jssc.SerialPortException;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.nio.ByteBuffer;
import java.util.function.Consumer;

import static java.lang.Thread.interrupted;
import static jssc.SerialPort.*;

public class FcReader {
    private SerialPort port;
    private Thread readerThread;
    private Consumer<Short> valueConsumer;

    public FcReader(final Consumer<Short> valueConsumer) throws SerialPortException, AWTException {
        this.valueConsumer = valueConsumer;
        port = new SerialPort("/dev/ttyUSB0");
        port.openPort();
        port.setParams(BAUDRATE_9600, DATABITS_8, STOPBITS_1, PARITY_NONE);

        readerThread = new Thread(new ReaderThread());
        readerThread.start();


        Robot robot = new Robot();
        new Thread(() -> {
            try {
                while (!Thread.interrupted()) {
                    robot.keyPress(KeyEvent.VK_RIGHT);
                    Thread.sleep(1500);
                    robot.keyRelease(KeyEvent.VK_RIGHT);
                    robot.keyPress(KeyEvent.VK_LEFT);
                    Thread.sleep(1500);
                    robot.keyRelease(KeyEvent.VK_LEFT);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void close() throws SerialPortException {
        port.closePort();
        readerThread.interrupt();
    }

    private class ReaderThread implements Runnable {
        @Override
        public void run() {
            try {
                while (!interrupted()) {
                    byte[] value =  port.readBytes(2);
                    byte upper = value[0];
                    byte lower = value[1];

                    if (isFc(upper)) {
                        if (isFc(lower)) {
                            continue;
                        } else {
                            upper = lower;
                            lower = port.readBytes(1)[0];
                        }
                    }
                    valueConsumer.accept(ByteBuffer.wrap(new byte[]{upper, lower}).getShort());
                    Thread.sleep(10);
                }
            } catch (SerialPortException | InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private boolean isFc(byte value) {
        int firstDigit = toDigit("FC".charAt(0));
        int secondDigit = toDigit("FC".charAt(1));
        return value == (byte) ((firstDigit << 4) + secondDigit);
    }

    private int toDigit(char hexChar) {
        int digit = Character.digit(hexChar, 16);
        if(digit == -1) {
            throw new IllegalArgumentException(
                    "Invalid Hexadecimal Character: "+ hexChar);
        }
        return digit;
    }
}

File: src/serial-keyboard/src/main/kotlin/serial/keyboard/serial/FcReader.kt
package serial.keyboard.serial

import com.fazecast.jSerialComm.SerialPort
import java.nio.ByteBuffer
import javax.inject.Singleton

@Singleton
class FcReader {
    fun parseByte(data: ByteArray, serialPort: SerialPort): Short {
        var upper: Byte = data[0]
        var lower: Byte = data[1]

        val singleByteBuffer = ByteArray(1)

        if (isFc(upper)) {
            if (isFc(lower)) {
                return -1;
            } else {
                serialPort.readBytes(singleByteBuffer, 1);

                upper = lower
                lower = singleByteBuffer[0]
            }
        }
        return ByteBuffer.wrap(byteArrayOf(upper, lower)).short
    }

    private fun isFc(value: Byte): Boolean {
        val firstDigit = toDigit("FC"[0])
        val secondDigit = toDigit("FC"[1])
        return value == ((firstDigit shl 4) + secondDigit).toByte()
    }

    private fun toDigit(hexChar: Char): Int {
        val digit = Character.digit(hexChar, 16)
        require(digit != -1) { "Invalid Hexadecimal Character: $hexChar" }
        return digit
    }
}